execution/order_by.go

```diff
@@ -18,132 +15,217 @@ const (
+	pref := OrderByKey{key: bytePref}
+
+	count  := octosql.MakePhantom()
+	err := recordCountMap.Get(&pref, &count)
+
+	if !record.IsUndo() {
+		if err == storage.ErrNotFound {
+
+			err = recordValueMap.Set(&pref, record)
 			if err != nil {
-				panic(errors.Wrap(err, "couldn't merge variables"))
+				return err
 			}
 
-			x, err := expr.ExpressionValue(ctx, iVars)
+			count = octosql.MakeInt(1)
+			err = recordCountMap.Set(&pref, &count)
 			if err != nil {
-				panic(errors.Wrapf(err, "couldn't get order by expression with index %v value", num))
+				return err
 			}
-			y, err := expr.ExpressionValue(ctx, jVars)
+		} else if err != nil {
+			return err
+		} else {
+			// Keep track of record vs retraction count
+			count = octosql.MakeInt(count.AsInt() + 1)
+			err = recordCountMap.Set(&pref, &count)
 			if err != nil {
-				panic(errors.Wrapf(err, "couldn't get order by expression with index %v value", num))
+				return err
 			}
-
-			if !isSorteable(x) {
-				panic(errors.Errorf("value %v of type %v is not comparable", x, reflect.TypeOf(x).String()))
-			}
-			if !isSorteable(y) {
-				panic(errors.Errorf("value %v of type %v is not comparable", y, reflect.TypeOf(y).String()))
+		}
+	} else {
+		if err == storage.ErrNotFound {
+			return errors.Wrap(err, "couldn't retract record")
+		} else if err != nil {
+			return err
+		} else {
+			// Keep track of record vs retraction count
+			count = octosql.MakeInt(count.AsInt() - 1)
+			if count.AsInt() < 0 {
+				return errors.New( "couldn't retract record")
 			}
 
-			cmp, err := octosql.Compare(x, y)
+			err = recordCountMap.Set(&pref, &count)
 			if err != nil {
-				panic(errors.Errorf("failed to compare values %v and %v", x, y))
+				return err
 			}
+		}
+	}
 
-			answer := false
+	return nil
+}
 
-			if cmp == 0 {
-				continue
-			} else if cmp > 0 {
-				answer = true
-			}
+func (ob *OrderByStream) Trigger(ctx context.Context, tx storage.StateTransaction, key octosql.Value) ([]*Record, error) {
+	output := make([]*Record, 0)
 
-			if directions[num] == Ascending {
-				answer = !answer
-			}
+	keyPrefix := append(append([]byte("$"), key.MonotonicMarshal()...), '$')
+	txByKey := tx.WithPrefix(keyPrefix)
+
+	var recordValueMap *storage.Map
+	var recordCountMap *storage.Map
 
-			return answer
+	recordValueMap = storage.NewMap(txByKey.WithPrefix(recordValuePrefix))
+	recordCountMap = storage.NewMap(txByKey.WithPrefix(recordCountPrefix))
+
+	var err error
+	pref := OrderByKey{key: []byte("")}
+	var count octosql.Value
+	var value Record
+	iter := recordValueMap.GetIterator()
+	for err = iter.Next(&pref, &value); err == nil; err = iter.Next(&pref, &value) {
+		err := recordCountMap.Get(&pref, &count)
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't get record count")
+		}
+
+		for i := 0; i < count.AsInt(); i++ {
+			newValue := value
+			output = append(output, &newValue)
```

Comment:
`newValue` is not needed here, just use `&value`

Reply:
it is needed because we need new reference for each value in output

Reply:
ok